### ISSUE TITLE
Hotkey behavior is inconsistent

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/geometric-optics#395]
**Test device**
Dell
**Operating System**
Win 11
**Browser**
Firefox
**Problem description**
For https://github.com/phetsims/qa/issues/787
When using J + P to jump the rulers to different points of interest I had trouble controlling this. The commands didn't always work and I wasn't able to determine what caused it to work vs. not work.

Troubleshooting information:

<details>

!!!!! DO NOT EDIT !!!!!
Name: ‪Geometric Optics‬
URL: https://phet-dev.colorado.edu/html/geometric-optics/1.1.0-rc.1/phet/geometric-optics_all_phet.html
Version: 1.1.0-rc.1 2022-03-18 20:41:59 UTC
Features missing: touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:98.0) Gecko/20100101 Firefox/98.0
Language: en-US
Window: 1280x667
Pixel Ratio: 1.5/1
WebGL: WebGL 1.0
GLSL: WebGL GLSL ES 1.0
Vendor: Mozilla (ANGLE (Intel, Intel(R) HD Graphics Direct3D11 vs_5_0 ps_5_0))
Vertex: attribs: 16 varying: 30 uniform: 4096
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 32767x32767
OES_texture_float: true
Dependencies JSON: {}

</details>

**Comment by pixelzoom** [comment]
Reproduced in master and 1.1.0-rc.1.   

In the above commits, I added logging to the hotkey callback - run master with `?log` and watch the console.  If the hotkey fires, you'll see "hotkey J+P" in the console.  What I'm seeing is that the J+P hotkey is not reliably firing. And it seems to fire less reliably if I release the J key before pressing the P key.  

Also very strange is that if I hold down both J + P, the hotkey fires repeatedly.

@jessegreenberg @zepumph is it possibly that multi-key hotkeys are broken again? And can you please remind me how they are they are supposed to be used?  Showing "J + P" in the keyboard help seems to imply that you need to press them both at the same time, like "Shift + Tab". But they (sometimes) work when you don't press them at the same time.

**Comment by pixelzoom** [comment]
I see similar flakey behavior with J+W and J+S in ballons-and-static-electricity master.  @KatieWoe can you confirm?

**Comment by pixelzoom** [comment]
This issue has made me re-think our hotkey choice for this action.  

When we started the design of rulers in https://github.com/phetsims/geometric-optics/issues/279, we originally had multiple hotkheys, one of each specific point. E.g. J+O, J+I, J+L, J+M, etc.  This is simlar to BASE, which has J+S, J+W, etc.

After interations on the design and implementation (e.g. https://github.com/phetsims/geometric-optics/issues/310) we've arrived at one hotkey that juump by cycling between points of interest.   The hotkey does not specify which specific.  

So since there is now only one hotkey for "jump", do we need a multi-key hotkey sequences? Can we simply use "J"?  It seems like it would be advantageous to avoid multi-key unless really needed, since a single key is easier to press.  And it would also allow this sim to side-step the implementation problems with mutli-key for now. 

@arouinfar what do you think?  Can use simply use "J" for our hotkey?

**Comment by KatieWoe** [comment]
I did see the difficulty on BASE in master, but not on BASE in the published sim. Hopefully that helps narrow some things down.

**Comment by arouinfar** [comment]
> @arouinfar what do you think? Can use simply use "J" for our hotkey?

This would be fine with me. Two-key hotkeys make sense when there are several unique options, but it's not needed for the current approach. 

> EDIT: It looks like the "J" was lost way back (late January?) when we switched from J+O, J+I, etc. to one hotkey that cycles through points. Very odd that it felt like it was working all this time.

Interesting! I missed it because I normally hold down J and then tap the other key as needed.

> But it looks like you currently have to wait some time between presses before the hotkey will fire again.

This is not my experience in rc.1 -- the hotkeys are really snappy.  Master is awful, however. I would consider the current behavior in master to be publication-blocking for any sim that utilizes hotkeys.

<details><summary>Troubleshooting Info for rc.1</summary>

Name: ‪Geometric Optics‬
URL: https://phet-dev.colorado.edu/html/geometric-optics/1.1.0-rc.1/phet/geometric-optics_en_phet.html
Version: 1.1.0-rc.1 2022-03-18 20:41:59 UTC
Features missing: applicationcache, applicationcache, touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.83 Safari/537.36
Language: en-US
Window: 1680x926
Pixel Ratio: 1.7999999523162842/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 31 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {}
</details>

**Comment by pixelzoom** [comment]
Unassigning @jessegreenberg and @zepumph because we're no longer using a multi-key hotkey.

In the above commits, I changed 'J+P' to simply 'J' in master and 1.1. @arouinfar please review in master, including the changes to the keyboard-help dialog.  If it looks OK, assign back to me.

I agree, hotkeys are very broken in master.  I've promoted that to https://github.com/phetsims/scenery/issues/1393, labeled blocking. I'm hopeful (but not certain) that the next 1.1 RC will be able to proceed.

**Comment by pixelzoom** [comment]
>  I'm hopeful (but not certain) that the next 1.1 RC will be able to proceed.

I just did a local build from the 1.1 branch.  Hotkeys feel very broken to me in 1.1 branch too.  For example, I can't just deliberately press the "J" key, I have to hold it down - for what feels like too long.  If I press and hold the "J" key, it fires repeately, which I guess is expected. The interval that it repeats at feels like its about the same as the amount of time I have to hold "J" down when I want it to fire once.  

So this is now blocking the next 1.1 RC. 

Assigning to @jessegreenberg and @zepumph for assistance.

**Comment by arouinfar** [comment]
@pixelzoom the updated keyboard help dialog looks good.

> For example, I can't just deliberately press the "J" key, I have to hold it down - for what feels like too long. If I press and hold the "J" key, it fires repeately, which I guess is expected. The interval that it repeats at feels like its about the same as the amount of time I have to hold "J" down when I want it to fire once.

I don't have a way to test the latest behavior in the 1.1 branch, but it sounds like my experience in master. The behavior in master truly awful right now. I completely agree that it's blocking.

**Comment by pixelzoom** [comment]
On hold until https://github.com/phetsims/scenery/issues/1355 is addressed.

**Comment by pixelzoom** [comment]
Patched and tested in 1.1 branch using this process:

```
% cd perennial
% grunt checkout-target --repo=geometric-optics --target=1.1
% cd scenery
% git checkout geometric-optics-1.1
% git cherry-pick bbddb3ccccb46c0826504f1bdffb7acac9084b75
% cd geometric-optics
% grunt 
// test built version
% cd scenery
% git push origin geometric-optics-1.1
% cd geometric-optics
% cp build/phet/dependencies.json .
% git add dependencies.json
% git push origin 1.1 
```

Ready for verification in next 1.1 RC.

**Comment by pixelzoom** [comment]
To verify in 1.1.0-rc.1:

* Go to PhET > Options, check "Add 2F checkbox to control panels"

For each screen:

* Check the "2F Points" checkbox in the control panel, so that the 2F points are visible.
* Use keyboard input to test each of the 4 tools.
* Use SPACE to move a tool out of the toolbox. It's initial position will be at the optic origin.
* Use ESCAPE to return a tool to the toolbox.
* Use the "J" hotkey (not J+P) to move a tool between points of interest. This will cycle through point from left-to-right, based on the dynamic position of the points, and whether the points are visible.
* Make some points invisible (e.g. uncheck "2F Points") to verify that invisible points are skipped

Points of interest vary by tool type, and are as follows:

horizontal ruler:
* origin is at its top-left corner
* the ruler's y coordinate is always on the optical axis
* the points of interest are:
    * object positions
    * image positions, if the image is to the left of the optic
    * optic position
    * F point to the left of the optic
    * 2F point to the left of the optic

vertical ruler:
* origin is at its top-left corner
* if the point to measure is on or above the optical axis, the ruler's y coordinate will be on the optical axis
* if the point to measure is below the optical axis, the ruler's y coordinate will be at the point of interest
* the points of interest are:
    * object positions
    * image positions
    * optic position
    * F points
    * 2F points
    * light spot position (for Light scene)
    * where rays converge to the right of the optic (for Light scene)

position markers:
* origin is at top-center
* the points of interest are:
    * object positions
    * image positions
    * optic position
    * light spot position (for Light scene)
    * where rays converge to the right of the optic (for Light scene)


If everything looks OK, please close this issue.

**Comment by KatieWoe** [comment]
Some inconsistencies with the point of interest:
For horizontal ruler, the object position is always on the dotted line (not clear if this is intended).
For vertical ruler, F and 2F are not included.
For position markers, F and 2F are included.

Reassigning since behavior does not match https://github.com/phetsims/geometric-optics/issues/395#issuecomment-1084875349. However, pressing J does consistently jump to a new point, so behavior is much better.

**Comment by pixelzoom** [comment]
@KatieWoe reported:
> For horizontal ruler, the object position is always on the dotted line (not clear if this is intended).

This behavior is correct, as noted in https://github.com/phetsims/geometric-optics/issues/395#issuecomment-1084875349 for horizontal ruler:
> * the ruler's y coordinate is always on the optical axis

@KatieWoe reported:
> For vertical ruler, F and 2F are not included. 
> For position markers, F and 2F are included.

This behavior is correct.  I think I must've accidentally put "F and 2F" under vertical ruler, when it should have been under "position markers".  I'll fix https://github.com/phetsims/geometric-optics/issues/395#issuecomment-1084875349.

So this can be closed.
